strongbox-metadata-maven-api/src/main/java/org/carlspring/strongbox/storage/metadata/MetadataMerger.java

```diff
@@ -108,19 +128,151 @@ public Metadata updateMetadataAtGroupLevel(PluginArtifact artifact, Metadata met
-            // TODO: Figure out how to get plugin name from artifact
-            plugin.setName("");
-            plugin.setArtifactId(artifact.getArtifactId());
-            // TODO: Figure out how to get plugin prefix from artifact
-            plugin.setPrefix("");
-
+            Plugin plugin = null;
+            plugin = getPluginFromArtifact(artifact);
             plugins.add(plugin);
         }
         return metadata;
     }
 
-    private Collection<SnapshotVersion> createNewSnapshotVersions(String version, String timestamp, int buildNumber)
+    /**
+     * This method will read the artifact and extract the plugin.xml from the artifact archive and extract it to
+     * temp folder in the same location as the artifact. It then reads plugin.xml using SAX parser and extracts information.
+     * After the information is read, the temp file and it's contents (plugin.xml) are deleted.
+     *
+     * @param artifact
+     * @return
+     */
+    private Plugin getPluginFromArtifact(PluginArtifact artifact)
+    {
+        Plugin plugin = new Plugin();
+        extractPluginXmlFromArtifact(artifact);
+        //Read the xml file using sax parser and set the plugin value for artifact and prefix
+        HashMap<String, String> pluginMap = readPluginXmlFileAndDelete(artifact);
+
+        plugin.setName(pluginMap.get("name"));
+        plugin.setPrefix(pluginMap.get("goalPrefix"));
+        plugin.setArtifactId(artifact.getArtifactId());
+        return plugin;
+    }
+
+    /**
+     * This method iterates over the archived artifact and calls the method extractPluginXmlFile() if the plugin.xml
+     * is found for extraction
+     *
+     * @param artifact
+     */
+    private void extractPluginXmlFromArtifact(PluginArtifact artifact)
+    {
+
+        FileInputStream fileInputStream = null;
+        ZipInputStream zipInputStream;
+        ZipEntry zipEntry;
+
+        try
+        {
+            fileInputStream = new FileInputStream(artifact.getFile().getCanonicalPath());
+            zipInputStream = new ZipInputStream(fileInputStream);
+
+            while ((zipEntry = zipInputStream.getNextEntry()) != null)
+            {
+                boolean isFileExtracted = extractPluginXmlFile(zipEntry, artifact, zipInputStream);
+                if (isFileExtracted)
+                {
+                    break;
+                }
+            }
+        }
+        catch (IOException e)
+        {
+            logger.error(
+                    "*** Error occurred while trying to extract plugin.xml from artifact " + artifact.getArtifactId() +
+                    " " + e.getMessage());
+        }
+
+    }
+
+    /**
+     * This method will extract plugin.xml from artifact archive and save it in the temp directory
+     *
+     * @param zipEntry
+     * @param artifact
+     * @param zipInputStream
+     * @return
+     * @throws IOException
+     */
+    private boolean extractPluginXmlFile(ZipEntry zipEntry,
```

Comment:
You should not be extracting this file. You can just grab it's `InputStream` from the zip.

Reply:
I thought of doing the same and keeping it in BufferedOutputStream but the problem with that approach is that if too many streams are running are running in parallel, it will eat up memory.